# **ML APPROACH**

- feature engineering for pooled gradient
boosting model

**LOADING AND CLEANING COMMODITY DATA**

In [47]:
# Libraries

from pathlib import Path
import pandas as pd
import numpy as np
from sklearn.model_selection import LeaveOneGroupOut

In [48]:
print(Path.cwd())

c:\Users\mason\OneDrive\Desktop\AgGenius\wasde_analysis\notebooks\models


In [49]:
PROJECT_ROOT = Path.cwd().parents[1]
DATA_DIR = PROJECT_ROOT / 'data' / 'processed' / 'commodity_prod'

print(Path.cwd())
print(DATA_DIR.exists())

c:\Users\mason\OneDrive\Desktop\AgGenius\wasde_analysis\notebooks\models
True


In [50]:
commodity_files = {
    'Corn': 'corn_prod.csv',
    'Cotton': 'cott_prod.csv',
    'Hard Red Spring Wheat': 'hrs_prod.csv',
    'Hard Red Winter Wheat': 'hrw_prod.csv',
    'Oats': 'oats_prod.csv',
    'Rice (Rough)': 'rr_prod.csv',
    'Soybean Meal': 'soy_meal_prod.csv',
    'Soybean Oil': 'soy_oil_prod.csv',
    'Soybeans': 'soybean_prod (2).csv',
    'Sugar': 'sugar_prod.csv',
    'Wheat': 'wht_prod.csv',
}

raw_data = {}
for commodity, filename in commodity_files.items():
    path = DATA_DIR / filename
    raw_data[commodity] = pd.read_csv(path, index_col=0, parse_dates=False)
    print(commodity, '->', raw_data[commodity].shape)

Corn -> (567, 15)
Cotton -> (567, 15)
Hard Red Spring Wheat -> (346, 15)
Hard Red Winter Wheat -> (346, 15)
Oats -> (567, 15)
Rice (Rough) -> (567, 15)
Soybean Meal -> (567, 15)
Soybean Oil -> (567, 15)
Soybeans -> (567, 15)
Sugar -> (567, 15)
Wheat -> (567, 15)


In [51]:
for commodity, d in raw_data.items():
    if 'ReliabilityProjection' in d.columns:
        d.drop(columns='ReliabilityProjection', inplace=True)
    if d['ReportDate'].dtype == str:
        d['ReportDate'] = pd.to_datetime(d['ReportDate'])

# Confirm the cleanup landed everywhere
for commodity, d in raw_data.items():
    print(commodity, d.shape, d['ReportDate'].dtype)

Corn (567, 14) str
Cotton (567, 14) str
Hard Red Spring Wheat (346, 14) str
Hard Red Winter Wheat (346, 14) str
Oats (567, 14) str
Rice (Rough) (567, 14) str
Soybean Meal (567, 14) str
Soybean Oil (567, 14) str
Soybeans (567, 14) str
Sugar (567, 14) str
Wheat (567, 14) str


In [52]:
print(f"Loaded {len(raw_data)} of {len(commodity_files)} expected commodities")
missing = set(commodity_files.keys()) - set(raw_data.keys())
print('Missing:', missing if missing else 'none')

Loaded 11 of 11 expected commodities
Missing: none


In [53]:
excluded_commodities = ['Hard Red Spring Wheat', 'Hard Red Winter Wheat']

raw_data_modeling = {k: v for k, v in raw_data.items() if k not in excluded_commodities}

print(f"Modeling on {len(raw_data_modeling)} commodities: {list(raw_data_modeling.keys())}")

Modeling on 9 commodities: ['Corn', 'Cotton', 'Oats', 'Rice (Rough)', 'Soybean Meal', 'Soybean Oil', 'Soybeans', 'Sugar', 'Wheat']


DateTime conversion

In [54]:
for commodity, d in raw_data_modeling.items():
    if not pd.api.types.is_datetime64_any_dtype(d['ReportDate']):
        d['ReportDate'] = pd.to_datetime(d['ReportDate'])
    print(commodity, d['ReportDate'].dtype)

Corn

 datetime64[us]
Cotton datetime64[us]
Oats datetime64[us]
Rice (Rough) datetime64[us]
Soybean Meal datetime64[us]
Soybean Oil datetime64[us]
Soybeans datetime64[us]
Sugar datetime64[us]
Wheat datetime64[us]


C:\Users\mason\AppData\Local\Temp\ipykernel_19116\2340286167.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  d['ReportDate'] = pd.to_datetime(d['ReportDate'])
C:\Users\mason\AppData\Local\Temp\ipykernel_19116\2340286167.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  d['ReportDate'] = pd.to_datetime(d['ReportDate'])
C:\Users\mason\AppData\Local\Temp\ipykernel_19116\2340286167.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  d['ReportDate'] = pd.to_datetime(d['ReportDate'])
C:\Users\mason\AppData\Local\Temp\ipykernel_19116\2340286167.py:3: UserWarning: Could not i

**Excluded:** Hard Red Spring Wheat, Hard Red Winter Wheat

WASDE never publishes a finalized (ProjEstFlag = NaN) production figure at the wheat-subclass level — confirmed across both raw commodity files and the master concatenated source, under every label variant. 

Only aggregate 'Wheat' production is ever finalized. Without ground truth, these two can't be validated as standalone nowcasting targets. 

**Possible future extension:** use HRS/HRW live estimates as input features for predicting total Wheat's finalized value, rather than as targets themselves

# **Run 'build_vintages' function across all commodities**

In [55]:
def build_vintages(df, commodity_name=None):
    """
    Given a cleaned WASDE dataframe for one commodity, build the vintage
    table: every live (Proj./Est.) estimate tagged with its horizon to
    finalization and the eventual final value.
    """
    df = df.copy()

    # First-finalized appearance per market year
    finals_info = (
        df[df['ProjEstFlag'].isna()]
          .sort_values('ReportDate')
          .groupby('MarketYear')
          .head(1)
          .set_index('MarketYear')[['ReportDate', 'Value']]
          .rename(columns={'ReportDate': 'final_date', 'Value': 'final_value'})
    )

    vintages = df[df['ProjEstFlag'].notna()].copy()
    vintages['final_value'] = vintages['MarketYear'].map(finals_info['final_value'])
    vintages['final_date'] = vintages['MarketYear'].map(finals_info['final_date'])

    vintages['horizon_months'] = (
        (vintages['final_date'].dt.year - vintages['ReportDate'].dt.year) * 12
        + (vintages['final_date'].dt.month - vintages['ReportDate'].dt.month)
    )

    # Drop any market year without a confirmed final value (still-live years)
    vintages = vintages.dropna(subset=['final_value', 'final_date']).copy()

    if commodity_name:
        vintages['Commodity'] = commodity_name

    return vintages, finals_info

In [56]:
all_vintages = []
finals_by_commodity = {}

for commodity, d in raw_data_modeling.items():
    v, finals_info = build_vintages(d, commodity_name=commodity)
    all_vintages.append(v)
    finals_by_commodity[commodity] = finals_info
    print(f"{commodity}: {len(v)} vintage rows, {v['MarketYear'].nunique()} finalized market years, max horizon {v['horizon_months'].max()}")

master_vintages = pd.concat(all_vintages, ignore_index=True)
print('\nTotal pooled rows:', master_vintages.shape)

Corn: 344 vintage rows, 16 finalized market years, max horizon 24.0
Cotton: 322 vintage rows, 16 finalized market years, max horizon 24.0
Oats: 344 vintage rows, 16 finalized market years, max horizon 24.0
Rice (Rough): 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Meal: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Oil: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybeans: 344 vintage rows, 16 finalized market years, max horizon 24.0
Sugar: 344 vintage rows, 16 finalized market years, max horizon 24.0
Wheat: 344 vintage rows, 16 finalized market years, max horizon 24.0

Total pooled rows: (3074, 17)


In [57]:
for commodity, d in raw_data_modeling.items():
    n_null_flag = d['ProjEstFlag'].isna().sum()
    unique_flags = d['ProjEstFlag'].unique()
    print(f"{commodity}: {n_null_flag} finalized rows | flags: {list(unique_flags)}")

Corn: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Cotton: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Oats: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Rice (Rough): 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Soybean Meal: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Soybean Oil: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Soybeans: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Sugar: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Wheat: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']


# **ESTABLISHING A BASELINE**

In [58]:
master_vintages['raw_error'] = master_vintages['final_value'] - master_vintages['Value']
master_vintages['raw_abs_error'] = master_vintages['raw_error'].abs()

master_vintages[['Commodity', 'MarketYear', 'horizon_months', 'Value', 'final_value', 'raw_error']].head(10)

,Commodity,MarketYear,horizon_months,Value,final_value,raw_error
0,Corn,2008/09,1.0,12092.0,12092.0,0.0
1,Corn,2009/10,13.0,13131.0,13092.0,-39.0
2,Corn,2009/10,12.0,13110.0,13092.0,-18.0
3,Corn,2010/11,24.0,13370.0,12447.0,-923.0
4,Corn,2009/10,11.0,13110.0,13092.0,-18.0
5,Corn,2010/11,23.0,13370.0,12447.0,-923.0
6,Corn,2009/10,10.0,13110.0,13092.0,-18.0
7,Corn,2010/11,22.0,13245.0,12447.0,-798.0
8,Corn,2009/10,9.0,13110.0,13092.0,-18.0
9,Corn,2010/11,21.0,13365.0,12447.0,-918.0


In [59]:
baseline_by_commodity_horizon = (
    master_vintages.groupby(['Commodity', 'horizon_months'])['raw_abs_error']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'mae_raw', 'count': 'n'})
    .reset_index()
)

print(baseline_by_commodity_horizon.head(20))

   Commodity  horizon_months     mae_raw   n
0       Corn             1.0    0.000000  16
1       Corn             2.0    0.000000  15
2       Corn             3.0    0.000000  15
3       Corn             4.0    0.000000  14
4       Corn             5.0   17.733333  15
5       Corn             6.0   17.733333  15
6       Corn             7.0   20.461538  13
7       Corn             8.0   26.466667  15
8       Corn             9.0   26.466667  15
9       Corn            10.0   26.466667  15
10      Corn            11.0   26.466667  15
11      Corn            12.0   29.133333  15
12      Corn            13.0   32.466667  15
13      Corn            14.0   32.000000  14
14      Corn            15.0   32.000000  14
15      Corn            16.0   28.307692  13
16      Corn            17.0  130.785714  14
17      Corn            18.0  130.785714  14
18      Corn            19.0  219.166667  12
19      Corn            20.0  259.000000  14


In [60]:
baseline_by_horizon_pooled = (
    master_vintages.groupby('horizon_months')['raw_abs_error']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'mae_raw', 'count': 'n'})
)

print(baseline_by_horizon_pooled)

                   mae_raw    n
horizon_months                 
1.0               0.152778  144
2.0               0.222222  135
3.0               1.600000  135
4.0               1.634921  126
5.0               3.911111  135
6.0               4.392593  135
7.0              28.237607  117
8.0              41.477612  134
9.0              66.149254  134
10.0             82.350746  134
11.0            105.485075  134
12.0            140.037313  134
13.0            168.506418  134
14.0            193.076960  125
15.0            187.557280  125
16.0            204.803793  116
17.0            226.640480  125
18.0            242.295680  125
19.0            294.460935  107
20.0            316.090000  124
21.0            317.107823  124
22.0            304.650484  124
23.0            363.614032  124
24.0            379.542258  124


In [61]:
master_vintages['raw_pct_error'] = (master_vintages['raw_error'] / master_vintages['final_value']).abs() * 100

baseline_by_commodity_horizon_full = (
    master_vintages.groupby(['Commodity', 'horizon_months'])
    .agg(
        mae_raw=('raw_abs_error', 'mean'),
        mape_raw=('raw_pct_error', 'mean'),
        n=('raw_abs_error', 'count')
    )
    .reset_index()
)

print(baseline_by_commodity_horizon_full.head(20))

   Commodity  horizon_months     mae_raw  mape_raw   n
0       Corn             1.0    0.000000  0.000000  16
1       Corn             2.0    0.000000  0.000000  15
2       Corn             3.0    0.000000  0.000000  15
3       Corn             4.0    0.000000  0.000000  14
4       Corn             5.0   17.733333  0.127743  15
5       Corn             6.0   17.733333  0.127743  15
6       Corn             7.0   20.461538  0.147396  13
7       Corn             8.0   26.466667  0.188648  15
8       Corn             9.0   26.466667  0.188648  15
9       Corn            10.0   26.466667  0.188648  15
10      Corn            11.0   26.466667  0.188648  15
11      Corn            12.0   29.133333  0.208227  15
12      Corn            13.0   32.466667  0.233116  15
13      Corn            14.0   32.000000  0.228489  14
14      Corn            15.0   32.000000  0.228489  14
15      Corn            16.0   28.307692  0.203151  13
16      Corn            17.0  130.785714  0.942466  14
17      Co

In [62]:
mape_by_commodity = (
    master_vintages.groupby('Commodity')['raw_pct_error']
    .mean()
    .sort_values()
)
print(mape_by_commodity)

Commodity
Wheat           0.777000
Corn            1.134595
Soybeans        1.331023
Rice (Rough)    1.429008
Soybean Meal    1.435916
Soybean Oil     1.447467
Sugar           1.618955
Cotton          2.561545
Oats            2.602168
Name: raw_pct_error, dtype: float64


- Wheat is the most predictable (0.78% average error) — USDA's live estimates for aggregate wheat production are already very close to final, across all horizons combined.
- Corn, Soybeans, Rice, Soybean Meal/Oil cluster in a middle band (1.1-1.4%) — decent, but with more room for a model to find real signal.
- Sugar (1.6%), Cotton (2.6%), Oats (2.6%) are the least predictable — WASDE's own early-season numbers for these are proportionally the roughest.

**LOCKED IN BASELINE**

In [63]:
baseline_by_commodity_horizon_full.to_csv(DATA_DIR / 'raw_wasde_baseline.csv', index=False)
print("Baseline saved.")

Baseline saved.


# **FEATURE ENGINEERING**

**Commodity**

In [64]:
master_vintages['Commodity'] = master_vintages['Commodity'].astype('category')
print(master_vintages['Commodity'].dtype)
print(master_vintages['Commodity'].cat.categories.tolist())

category
['Corn', 'Cotton', 'Oats', 'Rice (Rough)', 'Soybean Meal', 'Soybean Oil', 'Soybeans', 'Sugar', 'Wheat']


**Revision trajectory (how much the estimates moved from previous report)**

In [65]:
master_vintages = master_vintages.sort_values(['Commodity', 'MarketYear', 'ReportDate']).copy()

master_vintages['prior_value'] = (
    master_vintages.groupby(['Commodity', 'MarketYear'])['Value'].shift(1)
)

master_vintages['revision'] = master_vintages['Value'] - master_vintages['prior_value']
master_vintages['revision_pct'] = master_vintages['revision'] / master_vintages['prior_value'] * 100

master_vintages[['Commodity', 'MarketYear', 'ReportDate', 'Value', 'prior_value', 'revision', 'revision_pct']].head(15)

,Commodity,MarketYear,ReportDate,Value,prior_value,revision,revision_pct
0,Corn,2008/09,2010-04-01,12092.0,NaN,NaN,NaN
1,Corn,2009/10,2010-04-01,13131.0,NaN,NaN,NaN
2,Corn,2009/10,2010-05-01,13110.0,13131.0,-21.0,-0.159927
4,Corn,2009/10,2010-06-01,13110.0,13110.0,0.0,0.000000
6,Corn,2009/10,2010-07-01,13110.0,13110.0,0.0,0.000000
8,Corn,2009/10,2010-08-01,13110.0,13110.0,0.0,0.000000
11,Corn,2009/10,2010-09-01,13110.0,13110.0,0.0,0.000000
12,Corn,2009/10,2010-10-01,13110.0,13110.0,0.0,0.000000
14,Corn,2009/10,2010-11-01,13110.0,13110.0,0.0,0.000000
17,Corn,2009/10,2010-12-01,13110.0,13110.0,0.0,0.000000


In [66]:
print(master_vintages['revision'].isna().sum())
print(master_vintages['revision'].isna().sum() / len(master_vintages) * 100, '%')

144
4.68445022771633 %


In [67]:
master_vintages['is_first_estimate'] = master_vintages['prior_value'].isna().astype(int)
master_vintages['revision'] = master_vintages['revision'].fillna(0)
master_vintages['revision_pct'] = master_vintages['revision_pct'].fillna(0)

master_vintages[['Commodity', 'MarketYear', 'ReportDate', 'revision', 'revision_pct', 'is_first_estimate']].head(5)

,Commodity,MarketYear,ReportDate,revision,revision_pct,is_first_estimate
0,Corn,2008/09,2010-04-01,0.0,0.000000,1
1,Corn,2009/10,2010-04-01,0.0,0.000000,1
2,Corn,2009/10,2010-05-01,-21.0,-0.159927,0
4,Corn,2009/10,2010-06-01,0.0,0.000000,0
6,Corn,2009/10,2010-07-01,0.0,0.000000,0


**Report Month - Seasonality**

In [68]:
master_vintages['report_month'] = master_vintages['ReportDate'].dt.month

master_vintages.groupby('report_month').size().sort_index()

report_month
1     242
2     260
3     260
4     278
5     258
6     258
7     258
8     258
9     258
10    224
11    260
12    260
dtype: int64

In [69]:
master_vintages['report_month'] = master_vintages['report_month'].astype('category')

**Normalized version**

- Expresses each estimate relative to that commoditys own historical typical value

In [70]:
commodity_stats = (
    master_vintages.groupby('Commodity')['final_value']
    .agg(['mean', 'std'])
    .rename(columns={'mean': 'commodity_mean', 'std': 'commodity_std'})
)

master_vintages = master_vintages.merge(commodity_stats, on='Commodity', how='left')

master_vintages['value_zscore'] = (
    (master_vintages['Value'] - master_vintages['commodity_mean']) / master_vintages['commodity_std']
)

master_vintages[['Commodity', 'Value', 'commodity_mean', 'commodity_std', 'value_zscore']].groupby('Commodity').first()

,Value,commodity_mean,commodity_std,value_zscore
Commodity,,,,
Corn,12092.00,13764.758721,1184.069726,-1.412720
Cotton,12.82,16.106801,2.695367,-1.219426
Oats,89.00,63.433140,13.460932,1.899338
Rice (Rough),203.70,203.795640,22.465554,-0.004257
Soybean Meal,39104.00,46458.421512,4944.040844,-1.487533
Soybean Oil,18746.00,22813.078488,2671.251689,-1.522537
Soybeans,2967.00,3866.055233,498.922442,-1.801994
Sugar,7532.00,8801.069767,470.201137,-2.698993
Wheat,2499.00,1976.075581,212.129863,2.465115


What this does: 

for each commodity, compute the mean and standard deviation of its final (true) production values across all market years, 

then express every live estimate as a z-score relative to that — "how many standard deviations above/below this commodity's typical production level is this estimate." 

Now a value of 2.0 means the same thing (unusually high) whether it's Corn or Cotton, which the raw Value alone can't tell you.

**CONSOLIDATING THE FEATURE SET**

In [71]:
feature_cols = [
    'Commodity', 'horizon_months', 'revision', 'revision_pct',
    'is_first_estimate', 'report_month', 'value_zscore'
]
target_col = 'final_value'

print(master_vintages[feature_cols + [target_col]].isna().sum())
print()
print(master_vintages[feature_cols].dtypes)

Commodity            0
horizon_months       0
revision             0
revision_pct         0
is_first_estimate    0
report_month         0
value_zscore         0
final_value          0
dtype: int64

Commodity            category
horizon_months        float64
revision              float64
revision_pct          float64
is_first_estimate       int64
report_month         category
value_zscore          float64
dtype: object


In [72]:
feature_cols = [
    'Commodity', 'horizon_months', 'Value', 'value_zscore',
    'revision', 'revision_pct', 'is_first_estimate', 'report_month'
]

print(master_vintages[feature_cols].dtypes)

Commodity            category
horizon_months        float64
Value                 float64
value_zscore          float64
revision              float64
revision_pct          float64
is_first_estimate       int64
report_month         category
dtype: object


# **LEAVE-ONE-MARKET-YEAR-OUT SPLIT STRATEGY**

Since value_zscore depends on commodity_mean/commodity_std, and those were computed from final_value across all years, 

we need a version that recomputes them fold-by-fold using only the training years:

In [73]:
def add_zscore_feature(train_df, test_df):
    """
    Compute commodity mean/std from train_df only, then apply that
    same normalization to both train_df and test_df. Prevents the
    held-out year's own final_value from leaking into its own
    normalization.
    """
    stats = (
        train_df.groupby('Commodity')['final_value']
        .agg(['mean', 'std'])
        .rename(columns={'mean': 'commodity_mean', 'std': 'commodity_std'})
    )

    train_df = train_df.merge(stats, on='Commodity', how='left')
    test_df = test_df.merge(stats, on='Commodity', how='left')

    train_df['value_zscore'] = (train_df['Value'] - train_df['commodity_mean']) / train_df['commodity_std']
    test_df['value_zscore'] = (test_df['Value'] - test_df['commodity_mean']) / test_df['commodity_std']

    return train_df, test_df

In [74]:
cols_to_drop = ['commodity_mean', 'commodity_std', 'value_zscore']
master_vintages = master_vintages.drop(columns=[c for c in cols_to_drop if c in master_vintages.columns])

print(master_vintages.columns.tolist())

['ReportDate', 'ReportTitle', 'Attribute', 'Commodity', 'Region', 'MarketYear', 'ProjEstFlag', 'AnnualQuarterFlag', 'Value', 'Unit', 'ReleaseDate', 'ReleaseTime', 'ForecastYear', 'ForecastMonth', 'final_value', 'final_date', 'horizon_months', 'raw_error', 'raw_abs_error', 'raw_pct_error', 'prior_value', 'revision', 'revision_pct', 'is_first_estimate', 'report_month']


In [75]:
feature_cols_base = [
    'Commodity', 'horizon_months', 'Value', 'value_zscore',
    'revision', 'revision_pct', 'is_first_estimate', 'report_month'
]
target_col = 'final_value'

In [76]:
cotton = master_vintages[master_vintages['Commodity'] == 'Cotton']
print(sorted(cotton['MarketYear'].unique()))

overlap_check = cotton[cotton['MarketYear'].isin(['2010/11', '2010/011'])]
print(overlap_check[['MarketYear', 'ReportDate', 'Value']].sort_values('ReportDate'))

others = master_vintages[(master_vintages['MarketYear'].str.contains('/0')) & (master_vintages['Commodity'] != 'Cotton')]
print(others[['Commodity', 'MarketYear', 'ReportDate', 'Value']])

['2008/09', '2009/10', '2010/011', '2011/012', '2012/13', '2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23', '2023/24']
    MarketYear ReportDate  Value
358   2010/011 2010-05-01  16.70
359   2010/011 2010-06-01  16.70
360   2010/011 2010-07-01  18.30
361   2010/011 2010-08-01  18.53
362   2010/011 2010-09-01  18.84
363   2010/011 2010-10-01  18.87
364   2010/011 2010-11-01  18.42
365   2010/011 2010-12-01  18.27
366   2010/011 2011-01-01  18.32
367   2010/011 2011-02-01  18.32
368   2010/011 2011-03-01  18.32
369   2010/011 2011-04-01  18.10
370   2010/011 2011-05-01  18.10
371   2010/011 2011-06-01  18.10
372   2010/011 2011-07-01  18.10
373   2010/011 2011-08-01  18.10
374   2010/011 2011-09-01  18.10
375   2010/011 2011-10-01  18.10
376   2010/011 2011-11-01  18.10
377   2010/011 2011-12-01  18.10
378   2010/011 2012-01-01  18.10
379   2010/011 2012-02-01  18.10
380   2010/011 2012-03-01  18.10
381   2010/011 2012-04-01  18

In [77]:
# Fix Cotton's mislabeled market years at the source, before rebuilding vintages
raw_data_modeling['Cotton']['MarketYear'] = raw_data_modeling['Cotton']['MarketYear'].replace({
    '2010/011': '2010/11',
    '2011/012': '2011/12'
})

# Confirm the fix landed
print(sorted(raw_data_modeling['Cotton']['MarketYear'].unique()))

['2007/08', '2008/09', '2009/10', '2010/11', '2011/12', '2012/013', '2012/13', '2013/014', '2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23', '2023/24', '2024/25', '2025/26']


In [78]:
import re

cotton_years = sorted(raw_data_modeling['Cotton']['MarketYear'].unique())
malformed = [y for y in cotton_years if re.match(r'^\d{4}/0\d{2}$', y)]
print("Malformed years found:", malformed)

Malformed years found: ['2012/013', '2013/014']


In [79]:
def fix_market_year(y):
    m = re.match(r'^(\d{4})/0(\d{2})$', y)
    if m:
        return f"{m.group(1)}/{m.group(2)}"
    return y

raw_data_modeling['Cotton']['MarketYear'] = raw_data_modeling['Cotton']['MarketYear'].apply(fix_market_year)

# Confirm no malformed years remain
remaining = [y for y in raw_data_modeling['Cotton']['MarketYear'].unique() if re.match(r'^\d{4}/0\d{2}$', y)]
print("Remaining malformed:", remaining)
print(sorted(raw_data_modeling['Cotton']['MarketYear'].unique()))

Remaining malformed: []
['2007/08', '2008/09', '2009/10', '2010/11', '2011/12', '2012/13', '2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23', '2023/24', '2024/25', '2025/26']


In [80]:
all_vintages = []
finals_by_commodity = {}

for commodity, d in raw_data_modeling.items():
    v, finals_info = build_vintages(d, commodity_name=commodity)
    all_vintages.append(v)
    finals_by_commodity[commodity] = finals_info
    print(f"{commodity}: {len(v)} vintage rows, {v['MarketYear'].nunique()} finalized market years, max horizon {v['horizon_months'].max()}")

master_vintages = pd.concat(all_vintages, ignore_index=True)
print('\nTotal pooled rows:', master_vintages.shape)

Corn: 344 vintage rows, 16 finalized market years, max horizon 24.0
Cotton: 344 vintage rows, 16 finalized market years, max horizon 24.0
Oats: 344 vintage rows, 16 finalized market years, max horizon 24.0
Rice (Rough): 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Meal: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Oil: 344 vintage rows, 16 finalized market years, max horizon 24.0


Soybeans: 344 vintage rows, 16 finalized market years, max horizon 24.0
Sugar: 344 vintage rows, 16 finalized market years, max horizon 24.0
Wheat: 344 vintage rows, 16 finalized market years, max horizon 24.0

Total pooled rows: (3096, 17)


In [81]:
# Report month
master_vintages['report_month'] = master_vintages['ReportDate'].dt.month
master_vintages['report_month'] = master_vintages['report_month'].astype('category')

# Revision trajectory
master_vintages = master_vintages.sort_values(['Commodity', 'MarketYear', 'ReportDate']).copy()
master_vintages['prior_value'] = master_vintages.groupby(['Commodity', 'MarketYear'])['Value'].shift(1)
master_vintages['revision'] = master_vintages['Value'] - master_vintages['prior_value']
master_vintages['revision_pct'] = master_vintages['revision'] / master_vintages['prior_value'] * 100
master_vintages['is_first_estimate'] = master_vintages['prior_value'].isna().astype(int)
master_vintages['revision'] = master_vintages['revision'].fillna(0)
master_vintages['revision_pct'] = master_vintages['revision_pct'].fillna(0)

# Commodity as category
master_vintages['Commodity'] = master_vintages['Commodity'].astype('category')

# Raw + pct error vs baseline (for later comparison)
master_vintages['raw_error'] = master_vintages['final_value'] - master_vintages['Value']
master_vintages['raw_abs_error'] = master_vintages['raw_error'].abs()
master_vintages['raw_pct_error'] = (master_vintages['raw_error'] / master_vintages['final_value']).abs() * 100

print(master_vintages.shape)
print(master_vintages.isna().sum().sum(), 'total NaNs')

(3096, 25)
144 total NaNs


In [82]:
print(master_vintages[['revision', 'revision_pct']].isna().sum())
print(master_vintages['prior_value'].isna().sum())

revision        0
revision_pct    0
dtype: int64
144


In [83]:
cols_to_drop = ['commodity_mean', 'commodity_std', 'value_zscore']
master_vintages = master_vintages.drop(columns=[c for c in cols_to_drop if c in master_vintages.columns])

logo = LeaveOneGroupOut()
groups = master_vintages['MarketYear']

fold_results = []

for train_idx, test_idx in logo.split(master_vintages, groups=groups):
    train_raw = master_vintages.iloc[train_idx].copy()
    test_raw = master_vintages.iloc[test_idx].copy()

    train_fold, test_fold = add_zscore_feature(train_raw, test_raw)

    held_out_year = test_fold['MarketYear'].iloc[0]

    fold_results.append({
        'held_out_year': held_out_year,
        'n_train': len(train_fold),
        'n_test': len(test_fold)
    })

fold_summary = pd.DataFrame(fold_results)
print(fold_summary)
print('\nTotal folds:', len(fold_summary))

   held_out_year  n_train  n_test
0        2008/09     3087       9
1        2009/10     2979     117
2        2010/11     2880     216
3        2011/12     2880     216
4        2012/13     2889     207
5        2013/14     2889     207
6        2014/15     2880     216
7        2015/16     2880     216
8        2016/17     2880     216
9        2017/18     2889     207
10       2018/19     2889     207
11       2019/20     2880     216
12       2020/21     2880     216
13       2021/22     2889     207
14       2022/23     2889     207
15       2023/24     2880     216

Total folds: 16


# **FIT POOLED GRADIENT BOOSTING MODEL**

LightGBM or XGBoost on the full 3,074-row pooled set. Start simple (few features, shallow trees) before adding complexity.

# **EVALUATE AGAINST BASELINE**

Compare LOO MAE, pooled ML vs. raw WASDE estimate

# **RESULTS WRITE UP**